ZEP000-build-state-on-followers.md

```diff
@@ -0,0 +1,524 @@
+[comment]: <> (         - [ ] Does the ZEP break compatibility in `zeebe-gateway-protocol`?)
+[comment]: <> (         - [ ] Does the ZEP break compatibility in `zeebe-test`?)
+
+
+## Testing
+
+
+[comment]: <> (        You should describe what is the overall functionality that should be tested.)
+[comment]: <> (        If you are omitting tests, explain why, and explain the impact if it fails, specifically the worst case scenario.)
+[comment]: <> (        In each of the sections below, we should already list known cases that need to be tested in the final implementation, and at which level. The initial version here should be a best of effort: it is perfectly acceptable and expected that this section will be amended during implementation.)
+        
+### Unit
+
+### Integration
+
+  * We need to verify that the state is build on Leaders **AND** follower's. 
+  * We need to verify that the state is the SAME on Leaders **AND** follower's.
+    * Simple way would be to create an instance on the Leader, causing a Leader change and complete such an instance.
+    * We could run our randomized property based tests and verify the states.
+  * We need to make sure that the followers take snapshots and can compact their log.
+    * Verify the disk usage and whether snapshots have been taken, we already have QA tests for that.
+  * We need to verify that the blacklisting is build up correctly on the follower side and that if it becomes leader it ignores related commands.
+    * Either causing a blacklisting of an instance, triggering a Leader change and verify whether the instance is ignored or we take a look at the states.  
+  * In general, we should verify that fail-over works as expected and that the followers are building there state correctly, such that a new Leader can use it.
+     * Similar to above.
+  * After fail-over the new Leader should be able to export after the last lowest exporter position.
+  * After fail-over the new Leader should be able to process after the last processed command.
+  * InstallRequest's should be properly handled by followers, and they should continue with applying events after restoring the state.
+    * Maybe we verify whether new snapshots are taken, which is a good indication that it still makes progress.
+
+### E2E
+
+ In the E2E test we should verify similar things:
+
+ * We should verify that fail-over work as expected and that the followers are building there state correctly. The leaders should be able to continue where the followers left of and there should be no issue on exporting records from the new leader.
+
+### Benchmarks
+
+Verify the improvements in process execution latency during fail-over and also the latency from Follower-To-Leader processing transition, which was the main goal.
+ * New metric we should add here is the time between receiving the LEADER Role event and starting to process, because this is really the time we can decrease with the ZEP. 
```

Comment:
I don't see that as requirement but yes it would be interesting. I added it https://github.com/zeebe-io/enhancements/pull/16/commits/e8286c4ea83bb1992665444f04718a6be7456543